Add email preview functionality and refactor email handling

JumpLink opened the pull request:
This PR adds email preview functionality to the admin interface, allowing admins to see how emails will look before sending them. Users can now preview callout response emails and membership emails directly in the editor.

This PR is part of several pull requests:
* #349
* #351
* #355 (You are here)
* ~~#357~~
* #364
* #365

The changes also include a significant refactoring of the email handling code to improve maintainability and code organization. The [`EmailService`](packages/core/src/services/EmailService.ts) has been simplified, email templates have been extracted into a dedicated file for better reusability, and a new transformer pattern has been introduced for consistent email data handling across the application.

Can be tested on [dev](https://dev.beabee.io/) (latest update 28 Oct 17:40).

### Changes Made

#### Email Preview System
- **Server-side Rendering**: Added `EmailService.getTemplatePreview()` method for server-side email rendering with merge field resolution
- **Client-side Preview**: Enhanced `EmailEditor` component with live preview functionality
- **Preview API**: New email preview endpoint supporting all template types (general, contact, admin)
- **Merge Field Resolution**: Complete merge field processing including contact data, custom fields, and nested fields

#### Email Editor Improvements
- **Live Preview**: Real-time preview updates as users type in subject and content fields
- **Server Rendering**: Support for server-side rendered previews with proper merge field expansion
- **Loading States**: Loading indicators during server preview requests
- **Error Handling**: Fallback to client-side preview when server preview fails
- **Responsive Layout**: Side-by-side editor and preview panels with proper responsive behavior

#### Code Architecture Improvements
- **Email Template Extraction**: Moved email templates from inline code to dedicated data files
- **Type Safety**: Comprehensive TypeScript types for email editing and preview functionality
- **Utility Functions**: Extracted reusable functions for merge field processing and template handling
- **Email Preview Types**: New TypeScript interfaces for email preview data structures

#### Frontend Enhancements
- **Email Editor Component**: Major refactoring with support for multiple preview modes
- **Callout Email Integration**: Callout-specific email preview with proper merge field context
- **Membership Builder**: Updated email editor with improved preview capabilities
- **Router Updates**: Fixed typed router definitions for new email routes

#### Backend CLI Enhancements
- **Email Management Commands**: New CLI commands for managing email template overrides
- **Template Override System**: CRUD operations for custom email templates
- **Legacy Template Cleanup**: Migration to remove old confirmation templates

[Ticket](https://correctivdigital.openproject.com/projects/beabee/work_packages/1499)

wpf500 commented on apps/frontend/src/pages/admin/membership-builder.vue:
```diff
@@ -43,7 +43,7 @@ const tabs = computed(() =>
     },
     {
-      id: 'adminMembershipBuilderEmail',
+      id: '/admin/membership-builder/email',
```
Why this change?

JumpLink replied:
<img width="1393" height="409" alt="image" src="https://github.com/user-attachments/assets/ee4ec4c6-2379-496c-a3c7-3849bf286950" />
The generated route types have changed

wpf500 commented on apps/frontend/src/components/pages/admin/membership-builder/EmailEditor.vue:
```diff
@@ -1,66 +1,328 @@
-      .replace('*|FNAME|*', currentUser.value?.firstname || '')
-      .replace('*|LNAME|*', currentUser.value?.lastname || '') + props.footer
+// Use reactive reference to the template for two-way binding
+const template = reactive<EditableEmailTemplate>(
```
I think generally two-way data bindings are a bad idea unless the components are really tightly coupled. Here I think it would be better to use two `defineModel` props instead of a `template` prop. This would probably also clean up the `get`/`set` code you have in the parent file.

I think it would be something like this:

```ts
const subject = defineModel<string>('subject', {default: ''});
const content = defineModel<string>('content', {default: ''});
```

JumpLink replied:
Yes much better, thank you! Done with https://github.com/beabee-communityrm/monorepo/pull/355/commits/6c8daffe258d1afdd1ff44c7e781acfd2db3f8b3

wpf500 commented on apps/frontend/src/pages/admin/membership-builder/email.vue:
```diff
@@ -1,10 +1,3 @@
-<route lang="yaml">
```
Why remove this?

JumpLink commented on apps/frontend/src/components/pages/admin/membership-builder/EmailEditor.vue:
```diff
@@ -1,66 +1,328 @@
+);
+
+// Watch for content changes to update server preview
+watch(
```
Done with https://github.com/beabee-communityrm/monorepo/pull/355/commits/64307ce2a41ccf3cb7fab9923f599aaabd3a197c and seems to be working. The async `fetchServerPreview()` call is safely handled with `void` since it manages its own loading states.

In #357, I had implemented a debounce method for the preview. I am considering porting it back to this PR if the other one takes longer to be merged and would belittle the others somewhat

wpf500 commented on apps/backend/src/api/controllers/EmailController.ts:
```diff
@@ -61,7 +51,136 @@ export class EmailController {
+   */
+  @Authorized()
+  @Post('/preview/general/:templateId')
+  async previewGeneralEmail(
```
The three methods here are very repetitive, can they share a base? Or even the same controller method if `general` became a route placeholder

JumpLink replied:
I deliberately decided against a shared controller so that I could specify different controller parameters (like `@Authorized()` vs `@Authorized('admin')`) instead of checking this in the body. But I'm reworking it so that they share a common basis.

JumpLink commented:
I have now adjusted the code so that it is always rendered on the server side. For client-side rendering, I had also moved some email methods to common, which I have now reversed. I have also adopted the layout adjustments to the EmailEditor from #357, which means that the EmailEditor wraps better and always wraps in the Callout Editor, as there is simply not enough space here for a side-by-side view.
I could also remove the template in this branch, so that the next PR based on it really only contains the ANSWERS integration. There were originally several subsequent adjustments based on Sara's feedback, which is why it was included in #357.

JumpLink commented:
I have now adjusted the code so that it is always rendered on the server side. For client-side rendering, I had also moved some email methods to common, which I have now reversed. I have also adopted the layout adjustments to the EmailEditor from #357, which means that the EmailEditor wraps better.

I also backported the standard email templates - which are used when no callout response template is available - as this also fits with the preview feature in this PR. I will now split the PR for response summaries into two new PRs for:
* MergeFields button
* Response summaries

RayMille commented on packages/core/src/services/CalloutsService.ts:
```diff
@@ -335,13 +338,8 @@ class CalloutsService {
-          contact,
-          { calloutTitle: title, calloutSlug: callout.slug }
-        );
+        // Send confirmation email to the contact
```
It seems like this e-mail might be sent twice, because it is already called in this.setResponse() in function above. I do not think we need the duplication here (not tested).

JumpLink replied:
In `CalloutController.ts` a separate method is used for guests (this is not my change and initially confused me too):

```ts
    if (!caller || callout.access === 'only-anonymous') {
      const id = await calloutsService.setGuestResponse(
        callout,
        data.guest,
        data.answers,
        data.newsletter
      );
      return plainToInstance(GetGuestCalloutResponseDto, { id });
    } else {
      const response = await calloutsService.setResponse(
        callout,
        caller,
        data.answers,
        data.newsletter
      );
      return CalloutResponseTransformer.convert(response, auth, {});
    }
```

RayMille replied:
Hi @JumpLink, 

what I meant is that:

`await this.sendResponseEmail(callout, contact);`

is being sent twice in guest mode.

I tested it and this is happening:

<img width="325" height="246" alt="Screenshot from 2025-11-18 08-52-11" src="https://github.com/user-attachments/assets/2acb758b-0d42-4fce-bc41-96f1c57be609" />

JumpLink replied:
Fixed in the based branch `feat/callout-response-email-backend` with https://github.com/beabee-communityrm/monorepo/pull/351/commits/215e8712f84147358d2ab4573f0e1bf8fb76a838

RayMille commented on apps/frontend/src/components/pages/admin/membership-builder/EmailEditor.vue:
```diff
@@ -1,52 +1,211 @@
+          </div>
+
+          <!-- Preview content (server-rendered with footer and CSS) -->
+          <div v-else v-html="serverPreviewResult.body" />
```
Is this sanitized correctly? v-html is always a bit risky?

RayMille commented on packages/core/src/utils/email.ts:
```diff
@@ -8,3 +8,64 @@
+  mergeFields: Record<string, string>
+): string {
+  return Object.entries(mergeFields).reduce((result, [field, value]) => {
+    return result.replace(new RegExp(`\\*\\|${field}\\|\\*`, 'g'), value);
```
field names should be escaped for special characters

RayMille reviewed (commented):
Hi @JumpLink,

in general, the feature looks really nice, thank you for implementing. 

I do feel like we are creating a lot of extra boilerplate code. I am a bit worried that this might make the application really hard to read, debug in the long run.

Further, I would like to check the security implications of the server side rendering of the emails, and we need to test this feature and all emails before going live.

JumpLink commented on apps/frontend/src/components/pages/admin/callouts/tabs/ContentTab/SidebarTabContent/EmailTab.vue:
```diff
@@ -0,0 +1,161 @@
+watch(
+  emailData,
+  (newValue) => {
+    // eslint-disable-next-line vue/no-mutating-props
```
I tried using `defineModel()` instead, but it doesn't seem to work properly in this dynamic component setup with `<component :is>`. so I documented in https://github.com/beabee-communityrm/monorepo/pull/355/commits/8c199dd8c7004da2034e9ccdcbf971cd3e623f04 why this should be okay in this case.
